lxd/network/network_utils.go

```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
```

Comment:
I was now thinking a bit on this function and maybe we could pick a slightly different route that is easier to follow?

If we get the first and last address from the given `ipnet`, this will give us two byte slices (regardless of IPv4 or IPv6). Those IPs (byte slices) could be transferred into an `uint64` so that we can use a random number generator in between those min and max limits. The resulting number could then be translated back to either IPv4 or IPv6 (I guess this is the tricky part). 

Another approach would be to use the `net/netip.Addr.Next()` functions until a free one is found. But this wouldn't use any randomness.

Reply:
Thats why i was thinking to use https://pkg.go.dev/crypto/rand#Int

Reply:
But yes, your suggestion is basically the same as https://github.com/canonical/lxd/pull/13379#discussion_r1579197346 but mine was to take off the prefix and then calculate a random host number

Reply:
@roosterfish @markylaing what do you think of this:

```go
func randomIP(subnet net.IPNet) net.IP {
	// Take the starting IP of the subnet.
	startBig := big.NewInt(0)
	startBig.SetBytes(subnet.IP)

	// Calculate the host bits of the mask.
	ones, size := subnet.Mask.Size()
	hostBits := int64(size - ones)

	// Single host subnet.
	if hostBits <= 0 {
		return subnet.IP
	}

	// Calculate how many host IPs are available in the subnet.
	hostsBig := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(hostBits), nil)
	hostsBig.Sub(hostsBig, big.NewInt(1)) // Substract 1 so we don't go beyond the subnet.

	// Calculate the last IP in the subnet (optional, just for validation of logic).
	endBig := big.NewInt(0)
	endBig.Add(startBig, hostsBig)
	var endIP net.IP = endBig.Bytes()

	fmt.Printf("%v %v\n", subnet.IP.String(), endIP.String())

	// Calculate random IP in subnet.
	randBig, _ := rand.Int(rand.Reader, hostsBig)
	randBig.Add(startBig, randBig)
	var randIP net.IP = randBig.Bytes()

	return randIP
}
```

This seems to work well for both IPv4 and IPv6, including single host subnets.
What it doesn't cater for yet is avoiding returning the network or IPv4 broadcast addresses (there is no broadcast address in IPv6).

Reply:
I'm working on this currently and this is where I'm at:
```golang
// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
// each randomly generated value. Each time validate returns an error, another IP address will be generated until all
// IPs have been exhausted or the context is cancelled.
func randomAddressInSubnet(ctx context.Context, ipnet net.IPNet, validate func(net.IP) error) (net.IP, error) {
	ones, _ := ipnet.Mask.Size()

	// networkAddress is the first address in the subnet.
	var networkAddress net.IP
	var subnetExponent int64
	if ipnet.IP.To4() != nil {
		subnetExponent = 32 - int64(ones)

		// Apply mask to get first address in subnet. Use To4 to ensure the IP is in the correct form for use with big.Int.
		networkAddress = ipnet.IP.To4().Mask(ipnet.Mask)
	} else {
		subnetExponent = 128 - int64(ones)
		networkAddress = ipnet.IP.To16().Mask(ipnet.Mask)
	}

	subnetSize := big.NewInt(0).Exp(big.NewInt(2), big.NewInt(subnetExponent), nil)
	networkAddressInt := big.NewInt(0).SetBytes(networkAddress)

	// We want to exclude the network address and the broadcast address, so we'll exclude these from the range we generate
	// random numbers from.
	subnetSizeMinusTwo := big.NewInt(0).Sub(subnetSize, big.NewInt(2))

	// If the subnet is too small there are no addresses available. E.g. `192.0.2.0/31` contains only the network
	// address `192.0.2.0` and the broadcast address `192.0.2.1`.
	if subnetSizeMinusTwo.Cmp(big.NewInt(1)) <= 0 {
		return nil, noAvailableAddressErr{error: fmt.Errorf("No available addresses in subnet %q", ipnet.String())}
	}

	// Keep track of attempted IPs so that the validate function isn't repeatedly called in case it's expensive.
	// We don't need to worry about how large this map will grow. It is useful for small subnets where it's likely that
	// all IPs may be exhausted. The max size of a golang map will be the size of int, it is unfeasible for all of those
	// IP addresses to be in use (e.g. with `2001:db8::/32` the resource requirements for 2^32 IP addresses to be in use
	// precludes the possibility of this map growing too large).
	attempted := make(map[[16]byte]struct{})

	for {
		// Return on timeout or context cancellation.
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}

		// If we've attempted all possible addresses then return.
		if big.NewInt(int64(len(attempted))).Cmp(big.NewInt(0).Sub(subnetSizeMinusTwo, big.NewInt(1))) == 0 {
			return nil, noAvailableAddressErr{error: fmt.Errorf("No available addresses in subnet %q", ipnet.String())}
		}

		// Get a random number in range [0, subnetSizeMinusTwo).
		randBigInt, err := cryptorand.Int(cryptorand.Reader, subnetSizeMinusTwo)
		if err != nil {
			return nil, err
		}

		// Add the random number to the network address.
		randIPInt := big.NewInt(0).Add(networkAddressInt, randBigInt)

		// Add one to it to exclude the network address.
		randIP := net.IP(big.NewInt(0).Add(randIPInt, big.NewInt(1)).Bytes())
		randIP = randIP.To16()

		// Convert all IPs to a 16 byte representation so that we have a consistent map key we can use for both IPv4 and IPv6.
		var IPKey [16]byte
		copy(IPKey[:], randIP.To16())

		// Retry if we've already attempted this IP.
		_, ok := attempted[IPKey]
		if ok {
			continue
		}

		// Add the IP to the attempted map.
		attempted[IPKey] = struct{}{}

		// Call validate if set.
		if validate != nil {
			err = validate(randIP)
			if err != nil {
				continue
			}
		}

		return randIP, nil
	}
}
```

Reply:
It's passing all my unit tests (I've added edge cases and small subnets in there as well)

Reply:
That looks quite easy to follow and doesn't even need the end IP. 

But do we really need to check if it's a multicast address? Wouldn't it be already a misconfiguration on the network side if the functions gets passed a `net.IPNet` in one of the assigned multicast ranges?

Reply:
@roosterfish @markylaing updated my function now, seems to work well.

Reply:
@tomponline thanks, it was definitely useful to talk through this with you both earlier! I've updated my version with the changes we discussed and updated to unit tests accordingly. If you can think of any more edge cases to add to the unit tests that would be very helpful. Thanks.